### PR TITLE
Rounding bug with tax calculations

### PR DESCRIPTION
**Pull request by ciungulete**
Hi @sandervanhooft,

This pull request solve #224 

- Reproduce #224 with a test
- added `rounded_type` helper to see if amount need to be incremented or decremented
- add tests for this helper
- edit `FristPaymenPayment`:  `BaseAction`, `AddGenericOrderItem` and `FirstPaymentSubscriptionBuilder` to work with `rounded_type` helper

### CODE REVIEW

File: src/Helpers/helpers.php [hunk 1]
```diff
@@ -109,3 +109,28 @@ function money_to_decimal(Money $money)
         return (new DecimalMoneyFormatter(new ISOCurrencies()))->format($money);
     }
 }
+
+if (! function_exists('rounded_type')) {
```
Comment (sandervanhooft) [inline src/Helpers/helpers.php:113]: I think this should be in a separate (action) class or a method on `FirstPaymentSubscriptionBuilder`

**Review by sandervanhooft — CHANGES_REQUESTED** [review]
Thanks! Let's discuss this at 4PM.

The core issue is that we're doing the same calculation twice, but in a different sequence. This combined with rounding leads to different end results.

**Sequence 1 (FirstPayment)**

Actions -> Payment -> OrderItems -> Order

**Sequence 2 (mandated payment)**

OrderItems -> Order -> Payment

Your approach is a good starting point and technically correct. Definitely making modifications in the right place. Next step is to refactor this for clarity: what's going on and why we're doing it this way.

File: src/FirstPayment/Actions/AddGenericOrderItem.php [hunk 2]
```diff
@@ -13,16 +13,16 @@ class AddGenericOrderItem extends BaseAction
      * @param \Illuminate\Database\Eloquent\Model $owner
      * @param \Money\Money $subtotal
      * @param string $description
-     * @param int $roundedType
+     * @param int $roundedMode
```
Comment (sandervanhooft) [inline src/FirstPayment/Actions/AddGenericOrderItem.php:16]: Can you make it `roundingMode` so it's consistent with MoneyPHP?

File: src/FirstPayment/Actions/AddGenericOrderItem.php [hunk 3]
```diff
@@ -13,14 +13,16 @@ class AddGenericOrderItem extends BaseAction
      * @param \Illuminate\Database\Eloquent\Model $owner
      * @param \Money\Money $subtotal
      * @param string $description
+     * @param int $roundedType
```
[the remark this answers is not in the document]
Comment (ciungulete) [reply src/FirstPayment/Actions/AddGenericOrderItem.php]: Done!

File: src/Helpers/helpers.php [hunk 4]
```diff
@@ -109,3 +109,28 @@ function money_to_decimal(Money $money)
...
+
+    /**
+     * Format the money as basic decimal
+     *
+     * @param \Money\Money $total
+     * @param float $taxPercentage
```
[the remark this answers is not in the document]
Comment (ciungulete) [reply src/Helpers/helpers.php]: moved from there

File: src/FirstPayment/Actions/BaseAction.php [hunk 5]
```diff
@@ -25,6 +25,8 @@ abstract class BaseAction
     /** @var int */
     protected $quantity = 1;
 
+    /** @var int */
+    protected $roundedType = 5;
```
[the remark this answers is not in the document]
Comment (ciungulete) [reply src/FirstPayment/Actions/BaseAction.php]: Replace 5 with `Money::ROUND_HALF_UP`

File: src/FirstPayment/Actions/AddGenericOrderItem.php [hunk 6]
```diff
@@ -13,14 +13,16 @@ class AddGenericOrderItem extends BaseAction
...
      */
-    public function __construct(Model $owner, Money $subtotal, string $description)
+    public function __construct(Model $owner, Money $subtotal, string $description, int $roundedType = 5)
```
[the remark this answers is not in the document]
Comment (ciungulete) [reply src/FirstPayment/Actions/AddGenericOrderItem.php]: replace 5 with `Money::ROUND_HALF_UP`

File: src/FirstPayment/Actions/AddGenericOrderItem.php (hunk 2 above)
Comment (ciungulete) [reply src/FirstPayment/Actions/AddGenericOrderItem.php]: Renamed

**Comment by sandervanhooft** [comment]
Do we need the same on other Actions, like `AddBalance`?

**Comment by ciungulete** [comment]
> Do we need the same on other Actions, like `AddBalance`?

No, because `$this->taxPercentage = 0;` and we don't make any changes to `$subtotal`

**Review by sandervanhooft — APPROVED** [review]

**Pull request merged**